File: java/src/test/java/fr/sdecout/handson/rest/book/IsbnTest.java
package fr.sdecout.handson.rest.book;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;

public class IsbnTest {

    @Test
    void should_fail_to_initialize_from_value_with_less_than_13_digits() {
        var value = "978316148410";
        assertThat(value).hasSize(12);

        assertThatIllegalArgumentException()
                .isThrownBy(() -> new Isbn(value))
            .withMessage("ISBN must contain 13 numeric digits");
    }

    @Test
    void should_fail_to_initialize_from_value_with_more_than_13_digits() {
        var value = "97831614841000";
        assertThat(value).hasSize(14);

        assertThatIllegalArgumentException()
                .isThrownBy(() -> new Isbn(value))
                .withMessage("ISBN must contain 13 numeric digits");
    }

    @Test
    void should_fail_to_initialize_from_value_including_non_numeric_characters() {
        var value = "978316148410a";
        assertThat(value).hasSize(13);

        assertThatIllegalArgumentException()
                .isThrownBy(() -> new Isbn(value))
            .withMessage("ISBN must contain 13 numeric digits");
    }

    @Test
    void should_display_as_string() {
        assertThat(new Isbn("9783161484100")).hasToString("978-3-16-148410-0");
    }

}

File: kotlin/src/test/kotlin/fr/sdecout/handson/rest/book/IsbnTest.kt
package fr.sdecout.handson.rest.book

import fr.sdecout.handson.rest.book.Isbn
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatIllegalArgumentException
import org.junit.jupiter.api.Test

class IsbnTest {

    @Test
    fun `should fail to initialize from value with less than 13 digits`() {
        val value = "978316148410"
        assertThat(value).hasSize(12)

        assertThatIllegalArgumentException()
            .isThrownBy { Isbn(value) }
            .withMessage("ISBN must contain 13 numeric digits")
    }

    @Test
    fun `should fail to initialize from value with more than 13 digits`() {
        val value = "97831614841000"
        assertThat(value).hasSize(14)

        assertThatIllegalArgumentException()
            .isThrownBy { Isbn(value) }
            .withMessage("ISBN must contain 13 numeric digits")
    }

    @Test
    fun `should fail to initialize from value including non-numeric characters`() {
        val value = "978316148410a"
        assertThat(value).hasSize(13)

        assertThatIllegalArgumentException()
            .isThrownBy { Isbn(value) }
            .withMessage("ISBN must contain 13 numeric digits")
    }

    @Test
    fun `should display as string`() {
        assertThat(Isbn("9783161484100")).hasToString("978-3-16-148410-0")
    }

}
